---
title: "Prepare hgmm100 dataset"
author: "Lambda Moses"
date: "1/2/2019"
output: html_document
---

Below is how the 100 cell dataset are prepared. First, we process the fastq files to generate the gz file you download. Then we load the data from the gz file. The other datasets were prepared the same way.

## Download data
First download the fastq files from the 10x website
```{r}
download.file("http://cf.10xgenomics.com/samples/cell-exp/2.1.0/hgmm_100/hgmm_100_fastqs.tar", destfile = "hgmm_100_fastqs.tar")
```

Then untar this file
```{bash}
tar -xvf ./hgmm_100_fastqs.tar
```

## Build the `kallisto` index
Here we use [kallisto](https://pachterlab.github.io/kallisto/about) (see this link for install instructions) to pseudoalign the reads to the transcriptome and then to create the `bus` file to be converted to a sparse matrix. The first step is to build an index of the transcriptome. This data set has both human and mouse cells, so we need both human and mouse transcriptomes.
```{r, cache=TRUE}
# Human transcriptome
download.file("ftp://ftp.ensembl.org/pub/release-94/fasta/homo_sapiens/cdna/Homo_sapiens.GRCh38.cdna.all.fa.gz", "./hs_cdna.fa.gz", quiet = TRUE)
# Mouse transcriptome
download.file("ftp://ftp.ensembl.org/pub/release-94/fasta/mus_musculus/cdna/Mus_musculus.GRCm38.cdna.all.fa.gz", "./mm_cdna.fa.gz", quiet = TRUE)
```

Actually, we don't need to unzip the fasta files 
```{bash, cache=TRUE}
kallisto index -i hs_mm_tr_index.idx ./hs_cdna.fa.gz ./mm_cdna.fa.gz
```

## Run `kallisto bus`

```{bash}
kallisto bus -i ./hs_mm_tr_index.idx -o ./out_hgmm100 -x 10xv2 -t8 \
./fastqs/hgmm_100_S1_L001_R1_001.fastq.gz ./fastqs/hgmm_100_S1_L001_R2_001.fastq.gz \
./fastqs/hgmm_100_S1_L002_R1_001.fastq.gz ./fastqs/hgmm_100_S1_L002_R2_001.fastq.gz \
./fastqs/hgmm_100_S1_L003_R1_001.fastq.gz ./fastqs/hgmm_100_S1_L003_R2_001.fastq.gz \
./fastqs/hgmm_100_S1_L004_R1_001.fastq.gz ./fastqs/hgmm_100_S1_L004_R2_001.fastq.gz \
./fastqs/hgmm_100_S1_L005_R1_001.fastq.gz ./fastqs/hgmm_100_S1_L005_R2_001.fastq.gz \
./fastqs/hgmm_100_S1_L006_R1_001.fastq.gz ./fastqs/hgmm_100_S1_L006_R2_001.fastq.gz \
./fastqs/hgmm_100_S1_L007_R1_001.fastq.gz ./fastqs/hgmm_100_S1_L007_R2_001.fastq.gz \
./fastqs/hgmm_100_S1_L008_R1_001.fastq.gz ./fastqs/hgmm_100_S1_L008_R2_001.fastq.gz
```

See what are the outputs
```{r}
list.files("./out_hgmm100/")
```

## Run `BUStools`
The `output.bus` file is a binary. In order to make R parse it, we need to convert it into a sorted text file. There's a command line tool [`bustools`](https://github.com/BUStools/bustools) for this.

```{bash}
# Add where I installed bustools to PATH
export PATH=$PATH:/home/lambda/mylibs/bin/
# Sort
bustools sort -o ./out_hgmm100/output.sorted -t8 ./out_hgmm100/output.bus
# Convert sorted file to text
bustools text -o ./out_hgmm100/output.sorted.txt ./out_hgmm100/output.sorted
```

## Compress data for S3 storage
All files necessary to construct a sparse matrix with `BUSpaRse` are stored in Amazon S3 as a gz file. Here is how that gz file is made:
```{bash}
GZIP=-9 tar -cvzf hgmm100.tar.gz ./out_hgmm100/output.sorted.txt ./out_hgmm100/matrix.ec ./out_hgmm100/transcripts.txt
```

The gz file is what you download. Then it can be loaded with the `BUSpaRse` package. See the vignette of `BUSpaRse` for how to load the data downloaded from this package.
